R/data.table.R

```diff
@@ -3086,3 +3110,153 @@ isReallyReal = function(x) {
+  
+  if (missing(wt)){ 
+    if (ties) warning("Ties argument is TRUE without a wt column supplied. Ties argument will be ignored - head(dt, n) or, if n < 0, tail(dt, abs(n)) will be returned")
+    
+    if (n < 0) ind_sub = substitute(max(1, .N + n + 1):.N, list(n = n)) else ind_sub = substitute(1:min(n, .N), list(n = n)) 
+    
+    if (missing(by)){
+      return(list(substitute(ind <- inds, list(inds = ind_sub)))) 
+    } else {
+      l_args[['j']] = substitute(.I[inds], list(inds = ind_sub))
+    }
+  
+  } else {
+    
+    if (ties){
+      if (n < 0){
+        l_args[['j']] = substitute(.I[frank(wt, ties.method = 'min') <= n], list(n = abs(n), wt = substitute(wt)))
+      } else {
+        l_args[['j']] = substitute(.I[frank(-wt, ties.method = 'min') <= n], list(n = n, wt = substitute(wt)))
+      } 
+    } else {
+
+      i_j_subset = substitute(1:min(n, .N), list(n = abs(n)))
+      if (n < 0) i_order = substitute(order(wt), list(wt = substitute(wt))) else i_order = substitute(order(-wt), list(wt = substitute(wt)))
+      
+      if (missing(by)){
+        l_args[['i']] = substitute(order_i[i_subset], list(order_i = i_order, i_subset = i_j_subset))
+        l_args[['f_order']] = TRUE
+      } else {
+        l_args[['i']] = i_order
+        l_args[['j']] = substitute(.I[j_subset], list(j_subset = i_j_subset))
+      }
+    }
+  }
+
+  l_args
+}
+
+filter.at = function(cols, logic, by, all.vars = TRUE){
+  ## helper that takes takes filter.at() from the isub and massages it to produce indices for isub
+  
+  #' @param cols required; can accept all values that .SDcols accepts
+  #' @param logic required; a function or unquoted text that results in logic evaluation. The
+  #'              unquoted text must include `x` as an argument.
+  #'              If the argument contains `which.min` or `which.max`, intermediate results will
+  #'              be combined with `union` instead of `&` or `|`. 
+  #' @param by optional; accepts anything that the normal by argument accepts.
+  #' @param all.vars logical. Except as mentioned in `logic`, `all.vars = T` will reduce
+  #'                  results using `&` and `all.vars = F` will reduce results using `|`
+  #'                  
+  #' @return List. The number of list elements depends on the arguments above. 
+  #' 
+  #' `filter.at(cols = T, logic, all.vars = TRUE)` returns an unnamed list that needs to
+  #'      be evaulated in loop. This is to allow for some optimization of `dt[a & b & c]`.
+  #'      Argument can also be `cols = c('V1','V2')` to get the same  return as `cols = T`.
+  #' 
+  #' All other combinations return a named list to be evaluated as `do.call(`[data.table`, l_args)`
+  
+  cols = substitute(cols)
+  logic = substitute(logic)
+  
+  # for anonymous functions vs. functions in the environment
+  is_fx = any(as.character(logic) == 'x') || length(as.character(logic)) > 1
+
+  # Predicates wtih rle and cumsum do not work with chaining i.e. 
```

Comment:
what if there will be a user defined function that behaves like `rle`, `cumsum`, how we are going to handle that?

Reply:
The ```is_fx``` is poorly named. Mainly, the optimization is for if the user supplies an anonymous function. A previously defined function would be evaluated using ```Reduce```. That is, ```fx = cumsum(x) > 3; dt[filter.at(TRUE, fx)]``` would not be optimized in a for loop. However, ```dt[filter.at(TRUE, some_function(x))]```, would be optimized as long as ```some_function(x)``` is not ```'rle|cumsum|min|max|sum'```. TO DO: change ```is_fx``` to ```is_anon_fx``` for clarity but I am interested in what you are working on before resolving.